server/src-lib/Hasura/Eventing/ScheduledTrigger.hs

```diff
@@ -345,24 +379,44 @@ processCronEvents logger logEnv httpMgr pgpool getSC = do
                                        ctiComment
             finally <- runExceptT $
               runReaderT (processScheduledEvent logEnv pgpool scheduledEvent CronScheduledEvent) (logger, httpMgr)
+            removeEventFromLockedEvents id'
             either logInternalError pure finally
     Left err -> logInternalError err
   where
     logInternalError err = L.unLogger logger $ ScheduledTriggerInternalErr err
 
+    saveLockedCronEvents :: [CronEventPartial] -> IO ()
+    saveLockedCronEvents partialCronEvents =
+      liftIO $ atomically $ do
+        lockedEvents <- readTVar lockedCronEvents
+        let ids = map cepId partialCronEvents
+        writeTVar lockedCronEvents $!
+          Set.union lockedEvents $ Set.fromList ids
+
+    removeEventFromLockedEvents :: CronEventId -> IO ()
+    removeEventFromLockedEvents eventId =
+      liftIO $ atomically $ do
+        lockedEvents <- readTVar lockedCronEvents
+        writeTVar lockedCronEvents $! Set.delete eventId lockedEvents
```

Comment:
There’s quite a lot of duplicate logic between `processEventQueue`, `processCronEvents`, and `processStandAloneEvents`. Would it be possible to share more of that logic? It seems like maybe there’s a potential “event processor” abstraction here that could simplify a lot of this code!

Reply:
Yes, I have made a new module, which has common functions to share the locking/unlocking events logic